JavadocMethod: Javadoc Not Detected Above Multiline Comments

Check: http://checkstyle.sourceforge.net/config_javadoc.html#JavadocMethod

When programming in Java with JML annotations (http://www.jmlspecs.org/), the typical convention is to place the JML annotations - which are, for the purposes of any non-JML-aware tools, specially formatted Java comments - directly above the method they are annotating. 

A Javadoc comment also needs to be placed above the method it is documenting. Most Javadoc processing tools (including "javadoc" itself with the standard doclet) have no problem ignoring non-Javadoc comments that occur between the Javadoc comment for a method and the method declaration. However, Checkstyle will fail to detect the Javadoc comment for a method if there is a multi-line (but not a single-line) comment between the Javadoc comment and the method declaration. For example, Checkstyle correctly detects the Javadoc in this case:

```
/**
 * A Javadoc comment.
 */
//@ A JML Annotation
public void foo() {}
```

but not in this case:

```
/**
 * A Javadoc comment.
 */
/*@ A JML Annotation */
public void foo() {} 
```

It would be very helpful for Checkstyle to be able to detect such Javadoc comments. 

Examples of JML JavaDoc: https://sourceforge.net/apps/trac/jmlspecs/browser/TeachingMaterials/trunk/CAV2007Tutorial/examples/Actor.java?rev=2675 (https://sourceforge.net/apps/trac/jmlspecs/wiki/TeachingMaterials)
## 


This change impacts the following checks:
UnusedImports
JavadocMethod
JavadocStyle
JavadocType
JavadocVariable
MissingJavadocMethod
MissingJavadocType
WriteTag

Example of issue:

````
$ cat TestClass.java
public class TestClass {
/**
 * A Javadoc comment.
 */
//@ A JML Annotation
public int foo() { return 0; } 

/**
 * A Javadoc comment.
 */
/*@ A JML Annotation */
public int foo2() { return 0; } 
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="JavadocMethod" />
    </module>
</module>

$ java -jar checkstyle-8.27-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[ERROR] TestClass.java:6: Expected @return tag. [JavadocMethod]
Audit done.
Checkstyle ends with 1 errors.
````

Both methods should be handled the same but only 1 produces a violation.

As discussed above, issue is with `FileContents.getJavadocBefore` and this is used by 8 checks, so each one is affected by this bug. We have checks use this which are not AbstractJavadocCheck and probably will never be converted to it s it wouldn't make sense. Example: MissingJavadocMethodCheck .

So this issue needs to be fixed regardless of type of check it is. I believe we have other checks that use their own implementation of finding the javadoc.

--- a/src/main/java/com/puppycrawl/tools/checkstyle/api/FileContents.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/api/FileContents.java
@@ -216,12 +216,33 @@ public final class FileContents implements CommentListener {
         // Lines start at 1 to the callers perspective, so need to take off 2
         int lineNo = lineNoBefore - 2;
 
-        // skip blank lines
-        while (lineNo > 0 && (lineIsBlank(lineNo) || lineIsComment(lineNo))) {
+        // skip blank lines and comments
+        while (lineNo > 0 && (lineIsBlank(lineNo) || lineIsComment(lineNo)
+                            || lineInsideBlockComment(lineNo + 1))) {
             lineNo--;
         }
 
         return javadocComments.get(lineNo);
+    }
+
+    /**
+     * Checks if the specified line number is inside a block comment.
+     * This method scans through all block comments (excluding Javadoc comments)
+     * and determines whether the given line number falls within any of them
+     *
+     * @param lineNo the line number to check
+     * @return {@code true} if the line is inside a block comment (excluding Javadoc comments)
+     *          , {@code false} otherwise
+     */
+    private boolean lineInsideBlockComment(int lineNo) {
+        final Collection<List<TextBlock>> values = clangComments.values();
+        return values.stream()
+            .flatMap(List::stream)
+            .filter(comment -> !javadocComments.containsValue(comment))
+            .anyMatch(comment -> {
+                return lineNo >= comment.getStartLineNo()
+                                        && lineNo <= comment.getEndLineNo();
+            });
     }
 
     /**

--- a/src/test/java/com/puppycrawl/tools/checkstyle/api/FileContentsTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/api/FileContentsTest.java
@@ -202,11 +202,11 @@ public class FileContentsTest {
                 .that(fileContents.getJavadocBefore(4).toString())
                 .isEqualTo(new Comment(new String[] {"/** A */"}, 0, 1, 7).toString());
         assertWithMessage("Invalid comment")
-                .that(fileContents.getJavadocBefore(5))
-                .isNull();
+                .that(fileContents.getJavadocBefore(5).toString())
+                .isEqualTo(new Comment(new String[] {"/** A */"}, 0, 1, 7).toString());
         assertWithMessage("Invalid comment")
-                .that(fileContents.getJavadocBefore(6))
-                .isNull();
+                .that(fileContents.getJavadocBefore(6).toString())
+                .isEqualTo(new Comment(new String[] {"/** A */"}, 0, 1, 7).toString());
     }
 
     @Test

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheckTest.java
@@ -381,4 +381,13 @@ public class UnusedImportsCheckTest extends AbstractModuleTestSupport {
                 getPath("InputUnusedImportsFromStaticMethodRefInDocsOnly.java"), expected);
     }
 
+    @Test
+    public void testUnusedImportsJavadocAboveComments() throws Exception {
+        final String[] expected = {
+            "11:8: " + getCheckMessage(MSG_KEY, "java.util.List"),
+        };
+        verifyWithInlineConfigParser(
+                getPath("InputUnusedImportsJavadocAboveComments.java"), expected);
+    }
+
 }

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheckTest.java
@@ -593,4 +593,19 @@ public class JavadocMethodCheckTest extends AbstractModuleTestSupport {
         verifyWithInlineConfigParser(
                 getNonCompilablePath("InputJavadocMethodRecords3.java"), expected);
     }
+
+    @Test
+    public void testJavadocMethodAboveComments() throws Exception {
+        final String[] expected = {
+            "16:29: " + getCheckMessage(MSG_EXPECTED_TAG, "@throws", "Exception"),
+            "25:30: " + getCheckMessage(MSG_EXPECTED_TAG, "@throws", "Exception"),
+            "52:30: " + getCheckMessage(MSG_EXPECTED_TAG, "@throws", "Exception"),
+            "63:32: " + getCheckMessage(MSG_EXPECTED_TAG, "@throws", "Exception"),
+            "74:33: " + getCheckMessage(MSG_EXPECTED_TAG, "@throws", "Exception"),
+            "84:33: " + getCheckMessage(MSG_EXPECTED_TAG, "@throws", "Exception"),
+            "94:33: " + getCheckMessage(MSG_EXPECTED_TAG, "@throws", "Exception"),
+        };
+        verifyWithInlineConfigParser(
+                getPath("InputJavadocMethodAboveComments.java"), expected);
+    }
 }

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheckTest.java
@@ -822,4 +822,15 @@ public class JavadocStyleCheckTest
                 getPath("InputJavadocStyleCheck5.java"),
                 expected);
     }
+
+    @Test
+    public void testJavadocStyleAboveComments() throws Exception {
+        final String[] expected = {
+            "13: " + getCheckMessage(MSG_NO_PERIOD),
+        };
+
+        verifyWithInlineConfigParser(
+                getPath("InputJavadocStyleAboveComments.java"),
+                expected);
+    }
 }

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocTypeCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocTypeCheckTest.java
@@ -470,4 +470,13 @@ public class JavadocTypeCheckTest extends AbstractModuleTestSupport {
         verifyWithInlineConfigParser(
                 getPath("InputJavadocType4.java"), expected);
     }
+
+    @Test
+    public void testJavadocTypeAboveComments() throws Exception {
+        final String[] expected = {
+            "15:1: " + getCheckMessage(MSG_MISSING_TAG, "@author"),
+        };
+        verifyWithInlineConfigParser(
+                getPath("InputJavadocTypeAboveComments.java"), expected);
+    }
 }

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocVariableCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocVariableCheckTest.java
@@ -26,6 +26,7 @@ import org.junit.jupiter.api.Test;
 
 import com.puppycrawl.tools.checkstyle.AbstractModuleTestSupport;
 import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.CommonUtil;
 
 public class JavadocVariableCheckTest
     extends AbstractModuleTestSupport {
@@ -370,6 +371,14 @@ public class JavadocVariableCheckTest
         };
         verifyWithInlineConfigParser(
             getPath("InputJavadocVariableMethodInnerClass.java"),
+            expected);
+    }
+
+    @Test
+    public void testJavadocVariableAboveComment() throws Exception {
+        final String[] expected = CommonUtil.EMPTY_STRING_ARRAY;
+        verifyWithInlineConfigParser(
+            getPath("InputJavadocVariableAboveComment.java"),
             expected);
     }
 }

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocMethodCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocMethodCheckTest.java
@@ -528,4 +528,14 @@ public class MissingJavadocMethodCheckTest extends AbstractModuleTestSupport {
         verifyWithInlineConfigParser(
                 getNonCompilablePath("InputMissingJavadocMethodBasic.java"), expected);
     }
+
+    @Test
+    public void testMissingJavadocMethodAboveComments() throws Exception {
+        final String[] expected = {
+            "18:5: " + getCheckMessage(MSG_JAVADOC_MISSING),
+        };
+        verifyWithInlineConfigParser(
+                getPath("InputMissingJavadocMethodAboveComments.java"),
+                expected);
+    }
 }

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocTypeCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocTypeCheckTest.java
@@ -436,4 +436,14 @@ public class MissingJavadocTypeCheckTest extends AbstractModuleTestSupport {
             expected);
     }
 
+    @Test
+    public void testMissingJavadocTypeAboveComments() throws Exception {
+        final String[] expected = {
+            "13:1: " + getCheckMessage(MSG_JAVADOC_MISSING),
+        };
+        verifyWithInlineConfigParser(
+            getPath("InputMissingJavadocTypeAboveComments.java"),
+            expected);
+    }
+
 }

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/imports/unusedimports/InputUnusedImportsJavadocAboveComments.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/imports/unusedimports/InputUnusedImportsJavadocAboveComments.java
@@ -1,0 +1,20 @@
+/*
+UnusedImports
+processJavadoc = (default)true
+
+
+*/
+
+package com.puppycrawl.tools.checkstyle.checks.imports.unusedimports;
+
+import java.util.Map;
+import java.util.List; // violation 'Unused import - java.util.List.'
+
+/**
+ * Use {@link Map.Entry} in this javadoc.
+ */
+/*
+    Input for UnusedImports
+*/
+public class InputUnusedImportsJavadocAboveComments {
+}

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmethod/InputJavadocMethodAboveComments.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmethod/InputJavadocMethodAboveComments.java
@@ -1,0 +1,102 @@
+/*
+JavadocMethod
+validateThrows = true
+
+*/
+
+package com.puppycrawl.tools.checkstyle.checks.javadoc.javadocmethod;
+
+public class InputJavadocMethodAboveComments {
+
+    /**
+     * A Javadoc comment.
+     * @return 0
+     */
+    //@ A JML Annotation
+    public int foo() throws Exception { // violation, 'Expected @throws tag for 'Exception''
+        return 0;
+    }
+
+    /**
+     * A Javadoc comment.
+     * @return 0
+     */
+    /*@ A JML Annotation */
+    public int foo2() throws Exception { // violation, 'Expected @throws tag for 'Exception''
+        return 0;
+    }
+
+    /**
+     * A dangling Javadoc comment.
+     * @throws Exception exception
+     * @return 0
+     */
+    /**
+     * A Javadoc comment.
+     * @throws Exception exception
+     * @return 0
+     */
+    public int foo3() throws Exception {
+        return 0;
+    }
+
+    /**
+     * A dangling Javadoc comment.
+     * @throws Exception exception
+     * @return 0
+     */
+    /**
+     * A Javadoc comment.
+     * @return 0
+     */
+    public int foo4() throws Exception {  // violation, 'Expected @throws tag for 'Exception''
+        return 0;
+    }
+
+    /**
+     * A Javadoc comment.
+     * @return 0
+     */
+    @SuppressWarnings("unchecked")
+    /* generic type warnings are suppressed for this method
+    because of reason xyz */
+    public int method() throws Exception { // violation, 'Expected @throws tag for 'Exception''
+        return 0;
+    }
+
+    /**
+     * A Javadoc comment.
+     * @return 0
+     */
+    /* generic type warnings are suppressed for this method
+    because of reason xyz */
+    @SuppressWarnings("unchecked")
+    public int method2() throws Exception {  // violation,'Expected @throws tag for 'Exception''
+        return 0;
+    }
+
+    /**
+     * A Javadoc comment.
+     * @return 0
+     */
+    // generic type warnings are suppressed for this method because of reason xyz
+    @SuppressWarnings("unchecked")
+    public int method3() throws Exception { // violation, 'Expected @throws tag for 'Exception''
+        return 0;
+    }
+
+    /**
+     * A Javadoc comment.
+     * @return 0
+     */
+    @SuppressWarnings("unchecked")
+    // generic type warnings are suppressed for this method because of reason xyz */
+    public int method4() throws Exception { // violation, 'Expected @throws tag for 'Exception''
+        return 0;
+    }
+
+    /*@ A JML Annotation */
+    public int foo5() {
+        return 0;
+    }
+}

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocstyle/InputJavadocStyleAboveComments.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocstyle/InputJavadocStyleAboveComments.java
@@ -1,0 +1,18 @@
+/*
+JavadocStyle
+endOfSentenceFormat = (default)([.?!][ \t\n\r\f<])|([.?!]$)
+
+
+*/
+
+package com.puppycrawl.tools.checkstyle.checks.javadoc.javadocstyle;
+
+public class InputJavadocStyleAboveComments {
+
+     // violation below, 'First sentence should end with a period'
+     /**             Set of all class field names        */
+     /*
+        field
+     */
+    public String field;
+}

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadoctype/InputJavadocTypeAboveComments.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadoctype/InputJavadocTypeAboveComments.java
@@ -1,0 +1,28 @@
+/*
+JavadocType
+authorFormat = Mohamed Mahfouz
+
+
+*/
+
+
+package com.puppycrawl.tools.checkstyle.checks.javadoc.javadoctype;
+
+/** Test class for variable naming in for each clause. */
+/*
+   Input class for JavadocType
+*/
+public class InputJavadocTypeAboveComments {
+    // violation above, 'Type Javadoc comment is missing @author tag'
+}
+
+/**
+ * Test class for variable naming in for each clause.*
+ * @author Mohamed Mahfouz
+ */
+/*
+   Input class for JavadocType
+*/
+class MyClass {
+
+}

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocvariable/InputJavadocVariableAboveComment.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocvariable/InputJavadocVariableAboveComment.java
@@ -1,0 +1,16 @@
+/*
+JavadocVariable
+
+*/
+
+package com.puppycrawl.tools.checkstyle.checks.javadoc.javadocvariable;
+
+public class InputJavadocVariableAboveComment {
+    /**
+     * A public variable
+     */
+    /*
+       test comment
+     */
+    public int variablePublic;
+}

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/missingjavadocmethod/InputMissingJavadocMethodAboveComments.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/missingjavadocmethod/InputMissingJavadocMethodAboveComments.java
@@ -1,0 +1,28 @@
+/*
+MissingJavadocMethod
+minLineCount = (default)-1
+allowedAnnotations = (default)Override
+scope = private
+excludeScope = (default)null
+allowMissingPropertyJavadoc = (default)false
+ignoreMethodNamesRegex = (default)null
+tokens = (default)METHOD_DEF , CTOR_DEF , ANNOTATION_FIELD_DEF , COMPACT_CTOR_DEF
+
+
+*/
+
+package com.puppycrawl.tools.checkstyle.checks.javadoc.missingjavadocmethod;
+
+public class InputMissingJavadocMethodAboveComments {
+
+    public InputMissingJavadocMethodAboveComments() { }// violation
+
+    /**
+     *
+     */
+    /*
+     */
+    public InputMissingJavadocMethodAboveComments(String p1) {
+
+    }
+}

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/missingjavadoctype/InputMissingJavadocTypeAboveComments.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/missingjavadoctype/InputMissingJavadocTypeAboveComments.java
@@ -1,0 +1,22 @@
+/*
+MissingJavadocType
+scope = (default)public
+excludeScope = (default)null
+skipAnnotations = (default)Generated
+tokens = (default)INTERFACE_DEF, CLASS_DEF, ENUM_DEF, ANNOTATION_DEF, RECORD_DEF
+
+
+*/
+
+package com.puppycrawl.tools.checkstyle.checks.javadoc.missingjavadoctype;
+
+public class InputMissingJavadocTypeAboveComments { // violation, 'Missing a Javadoc comment.'
+}
+
+/**
+ *
+ */
+/* */
+class InputMissingJavadocTypeAboveComments2 {
+}
+

